File: test_all_options.py
#!/usr/bin/env python3
"""
Test all diffx options combinations for Python package
"""

import sys
import os
sys.path.insert(0, 'diffx-python/src')

from diffx.diffx import DiffOptions, diff, diff_string, DiffError
import tempfile
import json

# Test data
test_data1 = {
    "name": "Alice",
    "age": 30,
    "settings": {
        "theme": "dark",
        "notifications": True,
        "limits": {
            "max_files": 100,
            "timeout": 30.5
        }
    },
    "users": [
        {"id": 1, "name": "User1", "active": True},
        {"id": 2, "name": "User2", "active": False}
    ]
}

test_data2 = {
    "name": "Alice",
    "age": 31,
    "settings": {
        "theme": "light",
        "notifications": True,
        "limits": {
            "max_files": 150,
            "timeout": 30.501
        }
    },
    "users": [
        {"id": 1, "name": "User1", "active": True},
        {"id": 3, "name": "User3", "active": True}
    ]
}

def test_all_option_combinations():
    """Test all possible option combinations"""
    print("🧪 Testing all diffx option combinations...")
    
    # Create test files
    with tempfile.TemporaryDirectory() as tmp_dir:
        file1 = os.path.join(tmp_dir, "test1.json")
        file2 = os.path.join(tmp_dir, "test2.json")
        
        with open(file1, 'w') as f:
            json.dump(test_data1, f, indent=2)
        with open(file2, 'w') as f:
            json.dump(test_data2, f, indent=2)
        
        test_cases = [
            # Basic options
            ("Basic diff", DiffOptions()),
            ("JSON format", DiffOptions(format="json")),
            ("YAML format", DiffOptions(format="yaml")),
            # TOML format test skipped for JSON files - would need .toml files
            
            # Output formats
            ("CLI output", DiffOptions(output="cli")),
            ("JSON output", DiffOptions(output="json")),
            ("YAML output", DiffOptions(output="yaml")),
            ("Unified output", DiffOptions(output="unified")),
            
            # Path filtering
            ("Path filter", DiffOptions(path="settings.theme")),
            ("Nested path", DiffOptions(path="settings.limits.max_files")),
            
            # Regex ignore
            ("Ignore id keys", DiffOptions(ignore_keys_regex="^id$")),
            ("Ignore name keys", DiffOptions(ignore_keys_regex="name")),
            
            # Epsilon tolerance
            ("Epsilon 0.1", DiffOptions(epsilon=0.1)),
            ("Epsilon 0.001", DiffOptions(epsilon=0.001)),
            
            # Array ID key
            ("Array ID key", DiffOptions(array_id_key="id")),
            
            # Performance options
            ("Optimize", DiffOptions(optimize=True)),
            ("Batch size", DiffOptions(batch_size=500)),
            ("Optimize + Batch", DiffOptions(optimize=True, batch_size=500)),
            
            # Complex combinations
            ("Multi-option 1", DiffOptions(
                format="json",
                output="json",
                path="settings",
                epsilon=0.01
            )),
            ("Multi-option 2", DiffOptions(
                output="yaml",
                ignore_keys_regex="^id$",
                array_id_key="id",
                optimize=True
            )),
            ("Multi-option 3", DiffOptions(
                format="json",
                output="json",
                path="users",
                array_id_key="id",
                batch_size=100
            )),
            ("All options", DiffOptions(
                format="json",
                output="json",
                path="settings",
                ignore_keys_regex="timeout",
                epsilon=0.01,
                array_id_key="id",
                optimize=True,
                batch_size=1000
            ))
        ]
        
        success_count = 0
        total_count = len(test_cases)
        
        for name, options in test_cases:
            try:
                result = diff(file1, file2, options)
                print(f"✓ {name}: OK")
                success_count += 1
            except DiffError as e:
                print(f"✗ {name}: FAILED - {e}")
            except Exception as e:
                print(f"✗ {name}: ERROR - {e}")
        
        # Add TOML format test with proper files
        try:
            result = diff("tests/fixtures/file1.toml", "tests/fixtures/file2.toml", DiffOptions(format="toml"))
            print("✓ TOML format: OK")
            success_count += 1
        except DiffError as e:
            print(f"✗ TOML format: FAILED - {e}")
        except Exception as e:
            print(f"✗ TOML format: ERROR - {e}")
        
        total_count += 1
        
        print(f"\n📊 Results: {success_count}/{total_count} tests passed")
        return success_count == total_count

def test_string_diff_combinations():
    """Test string diff with various combinations"""
    print("\n🧪 Testing string diff combinations...")
    
    json1 = json.dumps(test_data1, indent=2)
    json2 = json.dumps(test_data2, indent=2)
    
    yaml1 = '''name: Alice
age: 30
settings:
  theme: dark
  notifications: true'''
    
    yaml2 = '''name: Alice
age: 31
settings:
  theme: light
  notifications: true'''
    
    test_cases = [
        ("String JSON", json1, json2, "json", DiffOptions()),
        ("String JSON with options", json1, json2, "json", DiffOptions(output="json", epsilon=0.1)),
        ("String YAML", yaml1, yaml2, "yaml", DiffOptions()),
        ("String YAML with options", yaml1, yaml2, "yaml", DiffOptions(output="yaml", path="settings")),
    ]
    
    success_count = 0
    total_count = len(test_cases)
    
    for name, content1, content2, format_type, options in test_cases:
        try:
            result = diff_string(content1, content2, format_type, options)
            print(f"✓ {name}: OK")
            success_count += 1
        except DiffError as e:
            print(f"✗ {name}: FAILED - {e}")
        except Exception as e:
            print(f"✗ {name}: ERROR - {e}")
    
    print(f"\n📊 String diff results: {success_count}/{total_count} tests passed")
    return success_count == total_count

def test_edge_cases():
    """Test edge cases and error conditions"""
    print("\n🧪 Testing edge cases...")
    
    test_cases = [
        ("Empty options", lambda: diff("tests/fixtures/config_v1.json", "tests/fixtures/config_v2.json", None)),
        ("Non-existent file", lambda: diff("nonexistent1.json", "nonexistent2.json", DiffOptions())),
        ("Invalid regex", lambda: diff("tests/fixtures/config_v1.json", "tests/fixtures/config_v2.json", 
                                      DiffOptions(ignore_keys_regex="[invalid"))),
    ]
    
    success_count = 0
    total_count = len(test_cases)
    
    for name, test_func in test_cases:
        try:
            if name == "Empty options":
                result = test_func()
                print(f"✓ {name}: OK (handled None options)")
                success_count += 1
            else:
                result = test_func()
                print(f"✗ {name}: Should have failed but didn't")
        except DiffError as e:
            print(f"✓ {name}: OK (properly caught error)")
            success_count += 1
        except Exception as e:
            print(f"✗ {name}: Unexpected error - {e}")
    
    print(f"\n📊 Edge case results: {success_count}/{total_count} tests passed")
    return success_count == total_count

if __name__ == "__main__":
    print("🔬 diffx Python Package - Comprehensive Option Testing")
    print("=" * 60)
    
    all_passed = True
    
    # Test all option combinations
    all_passed &= test_all_option_combinations()
    
    # Test string diff combinations  
    all_passed &= test_string_diff_combinations()
    
    # Test edge cases
    all_passed &= test_edge_cases()
    
    print("\n" + "=" * 60)
    if all_passed:
        print("🎉 All tests passed! Python package handles all option combinations correctly.")
        sys.exit(0)
    else:
        print("❌ Some tests failed. Please check the output above.")
        sys.exit(1)

File: test_all_options.js
#!/usr/bin/env node
/**
 * Test all diffx options combinations for npm package
 */

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Test data
const testData1 = {
    name: "Alice",
    age: 30,
    settings: {
        theme: "dark",
        notifications: true,
        limits: {
            max_files: 100,
            timeout: 30.5
        }
    },
    users: [
        { id: 1, name: "User1", active: true },
        { id: 2, name: "User2", active: false }
    ]
};

const testData2 = {
    name: "Alice",
    age: 31,
    settings: {
        theme: "light",
        notifications: true,
        limits: {
            max_files: 150,
            timeout: 30.501
        }
    },
    users: [
        { id: 1, name: "User1", active: true },
        { id: 3, name: "User3", active: true }
    ]
};

function runDiffx(args) {
    return new Promise((resolve, reject) => {
        const binaryPath = path.join(__dirname, 'diffx-npm', 'bin', 'diffx');
        const child = spawn(binaryPath, args, { stdio: 'pipe' });
        
        let stdout = '';
        let stderr = '';
        
        child.stdout.on('data', (data) => {
            stdout += data.toString();
        });
        
        child.stderr.on('data', (data) => {
            stderr += data.toString();
        });
        
        child.on('close', (code) => {
            if (code === 0) {
                resolve({ stdout, stderr });
            } else {
                reject({ code, stdout, stderr });
            }
        });
        
        child.on('error', (err) => {
            reject({ error: err.message });
        });
    });
}

async function testAllOptionCombinations() {
    console.log("🧪 Testing all diffx option combinations...");
    
    // Create temporary files
    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'diffx-test-'));
    const file1 = path.join(tmpDir, 'test1.json');
    const file2 = path.join(tmpDir, 'test2.json');
    
    fs.writeFileSync(file1, JSON.stringify(testData1, null, 2));
    fs.writeFileSync(file2, JSON.stringify(testData2, null, 2));
    
    const testCases = [
        // Basic options
        { name: "Basic diff", args: [file1, file2] },
        { name: "JSON format", args: [file1, file2, "--format", "json"] },
        { name: "YAML format", args: [file1, file2, "--format", "yaml"] },
        // TOML format test skipped for JSON files - would need .toml files
        
        // Output formats
        { name: "CLI output", args: [file1, file2, "--output", "cli"] },
        { name: "JSON output", args: [file1, file2, "--output", "json"] },
        { name: "YAML output", args: [file1, file2, "--output", "yaml"] },
        { name: "Unified output", args: [file1, file2, "--output", "unified"] },
        
        // Path filtering
        { name: "Path filter", args: [file1, file2, "--path", "settings.theme"] },
        { name: "Nested path", args: [file1, file2, "--path", "settings.limits.max_files"] },
        
        // Regex ignore
        { name: "Ignore id keys", args: [file1, file2, "--ignore-keys-regex", "^id$"] },
        { name: "Ignore name keys", args: [file1, file2, "--ignore-keys-regex", "name"] },
        
        // Epsilon tolerance
        { name: "Epsilon 0.1", args: [file1, file2, "--epsilon", "0.1"] },
        { name: "Epsilon 0.001", args: [file1, file2, "--epsilon", "0.001"] },
        
        // Array ID key
        { name: "Array ID key", args: [file1, file2, "--array-id-key", "id"] },
        
        // Performance options
        { name: "Optimize", args: [file1, file2, "--optimize"] },
        { name: "Batch size", args: [file1, file2, "--batch-size", "500"] },
        { name: "Optimize + Batch", args: [file1, file2, "--optimize", "--batch-size", "500"] },
        
        // Complex combinations
        { name: "Multi-option 1", args: [file1, file2, "--format", "json", "--output", "json", "--path", "settings", "--epsilon", "0.01"] },
        { name: "Multi-option 2", args: [file1, file2, "--output", "yaml", "--ignore-keys-regex", "^id$", "--array-id-key", "id", "--optimize"] },
        { name: "Multi-option 3", args: [file1, file2, "--format", "json", "--output", "json", "--path", "users", "--array-id-key", "id", "--batch-size", "100"] },
        { name: "All options", args: [file1, file2, "--format", "json", "--output", "json", "--path", "settings", "--ignore-keys-regex", "timeout", "--epsilon", "0.01", "--array-id-key", "id", "--optimize", "--batch-size", "1000"] }
    ];
    
    let successCount = 0;
    let totalCount = testCases.length;
    
    for (const testCase of testCases) {
        try {
            const result = await runDiffx(testCase.args);
            console.log(`✓ ${testCase.name}: OK`);
            successCount++;
        } catch (error) {
            console.log(`✗ ${testCase.name}: FAILED - ${error.stderr || error.error || error.code}`);
        }
    }
    
    // Add TOML format test with proper files
    try {
        const result = await runDiffx(["tests/fixtures/file1.toml", "tests/fixtures/file2.toml", "--format", "toml"]);
        console.log("✓ TOML format: OK");
        successCount++;
    } catch (error) {
        console.log(`✗ TOML format: FAILED - ${error.stderr || error.error || error.code}`);
    }
    totalCount++;
    
    // Clean up
    fs.rmSync(tmpDir, { recursive: true, force: true });
    
    console.log(`\n📊 Results: ${successCount}/${totalCount} tests passed`);
    return successCount === totalCount;
}

async function testDirectoryComparison() {
    console.log("\n🧪 Testing directory comparison...");
    
    const testCases = [
        { name: "Directory recursive", args: ["tests/fixtures/dir1", "tests/fixtures/dir2", "--recursive"] },
        { name: "Directory with format", args: ["tests/fixtures/dir1", "tests/fixtures/dir2", "--recursive", "--format", "json"] },
        { name: "Directory with output", args: ["tests/fixtures/dir1", "tests/fixtures/dir2", "--recursive", "--output", "json"] },
        { name: "Directory with path", args: ["tests/fixtures/dir1", "tests/fixtures/dir2", "--recursive", "--path", "name"] },
        { name: "Directory with all options", args: ["tests/fixtures/dir1", "tests/fixtures/dir2", "--recursive", "--format", "json", "--output", "json", "--optimize", "--batch-size", "100"] }
    ];
    
    let successCount = 0;
    let totalCount = testCases.length;
    
    for (const testCase of testCases) {
        try {
            const result = await runDiffx(testCase.args);
            console.log(`✓ ${testCase.name}: OK`);
            successCount++;
        } catch (error) {
            console.log(`✗ ${testCase.name}: FAILED - ${error.stderr || error.error || error.code}`);
        }
    }
    
    console.log(`\n📊 Directory comparison results: ${successCount}/${totalCount} tests passed`);
    return successCount === totalCount;
}

async function testEdgeCases() {
    console.log("\n🧪 Testing edge cases...");
    
    const testCases = [
        { name: "Non-existent file", args: ["nonexistent1.json", "nonexistent2.json"] },
        { name: "Invalid regex", args: ["tests/fixtures/config_v1.json", "tests/fixtures/config_v2.json", "--ignore-keys-regex", "[invalid"] },
        { name: "Invalid epsilon", args: ["tests/fixtures/config_v1.json", "tests/fixtures/config_v2.json", "--epsilon", "invalid"] },
        { name: "Invalid batch size", args: ["tests/fixtures/config_v1.json", "tests/fixtures/config_v2.json", "--batch-size", "invalid"] },
    ];
    
    let successCount = 0;
    let totalCount = testCases.length;
    
    for (const testCase of testCases) {
        try {
            const result = await runDiffx(testCase.args);
            console.log(`✗ ${testCase.name}: Should have failed but didn't`);
        } catch (error) {
            console.log(`✓ ${testCase.name}: OK (properly caught error)`);
            successCount++;
        }
    }
    
    console.log(`\n📊 Edge case results: ${successCount}/${totalCount} tests passed`);
    return successCount === totalCount;
}

async function main() {
    console.log("🔬 diffx npm Package - Comprehensive Option Testing");
    console.log("=".repeat(60));
    
    let allPassed = true;
    
    // Test all option combinations
    allPassed &= await testAllOptionCombinations();
    
    // Test directory comparison
    allPassed &= await testDirectoryComparison();
    
    // Test edge cases
    allPassed &= await testEdgeCases();
    
    console.log("\n" + "=".repeat(60));
    if (allPassed) {
        console.log("🎉 All tests passed! npm package handles all option combinations correctly.");
        process.exit(0);
    } else {
        console.log("❌ Some tests failed. Please check the output above.");
        process.exit(1);
    }
}

main().catch(console.error);